app/widget/timeruler/seekablewidget.cpp

```diff
@@ -127,6 +133,23 @@ void SeekableWidget::SetScroll(int s)
   update();
 }
 
+void SeekableWidget::UpdateMarkers()
+{
+  printf("Test");
+  if (timeline_points()) {
+    foreach(Marker * marker, marker_widgets) {
+      marker->deleteLater();
+    }
+    marker_widgets.clear();
+   
+    foreach(TimelineMarker * marker, timeline_points()->markers()->list()) {
+      Marker* marker_widget = new Marker();
+      marker_widgets.append(marker_widget);
+      marker_layout_->addWidget(marker_widget);
+    }
+  }
+}
+
```

Comment:
Would probably be inclined to have separate `MarkerAdd` and `MarkerRemove` functions that just work with one marker at a time, rather than destroying and creating them all each time. I often use a QMap or QHash to match source objects with their widgets (see NodeViewScene::item_map_)

Reply:
I've added a basic add marker function. How owuld you want to delete a marker? I suppose there needs to be some UI indication that it's been selected. Would you use the delete key or right click->delete?

Reply:
I think you should add the "edit menu items" (from MenuShared) to the context menu, which will add things like undo, redo, delete, cut, copy, paste, etc. Then add some logic for TimeBasedWidget to pass through from functions like `DeleteSelected` to the TimeRuler. Functions like "HandPress" and "PlayheadPress" might provide a good reference for implementing that. Doing that will ensure both right click > delete and the delete key will work.

Reply:
So I'm going to have to implement a bunch of AddMarkerUndo Command type classes then as well?

If I copy a marker, I assume all I'm copying is the colour and label? You wouldn't want the time/position data would you?